Complete first section of Pauli Twirling user guide (#2454)
unitaryfund/mitiq: docs/source/guide/pt-1-intro.md

# How do I use PT?

```{admonition} Warning:

> This users guide is still under construction and may change in the near future
> after some utility functions are introduced. 

Reviewer: I know you didn't write this, but lets keep it simple :)
```suggestion
This users guide is still under construction and may change in the near future.
```

mitiq.SUPPORTED_PROGRAM_TYPES.keys()
```

In this first section, we see how to use PT in Mitiq, starting from a circuit of interest.

+++

## Problem setup

> We first define the circuit, which in this example contains Hadamard, C-NOT, and C-Z gates.

Reviewer 2: ```suggestion
We first define the circuit, which in this example contains Hadamard (H), CNOT, and CZ gates.
```

state. See the [Executors](executors.md) section for more information on
how to define more advanced executors.

As per the noise model that is applied during execution by the simulator, 
we choose a depolarizing channel applied to the output of each 2-qubit gates.
Admittedly, this is not necessarily the most realistic noise,
but for the sake of this tutorial, it is going to be useful for highlighting 
the effect of Pauli Twirling.  

```{code-cell} ipython3
from cirq import CZPowGate, CXPowGate, CircuitOperation, depolarize, DensityMatrixSimulator
from cirq.devices.noise_model import GateSubstitutionNoiseModel

def get_noise_model(noise_level: float) -> GateSubstitutionNoiseModel:
    """Substitute each C-Z and C-NOT gate in the circuit 
    with the gate itself followed by a depolarizing channel
    """
    def noisy_c_gate(op):
        if isinstance(op.gate, (CZPowGate, CXPowGate)):
            return CircuitOperation(
                Circuit(
                    op.gate.on(*op.qubits), 

>                     depolarize(p=noise_level, n_qubits=2).on_each(op.qubits)

Reviewer 2: What if you change `depolarize` to a coherent noise channel? As we can describe a depolarizing channel using [paulis](https://quantumcomputing.stackexchange.com/questions/4978/how-to-find-the-operator-sum-representation-of-the-depolarizing-channel), I don't think pauli twirling is going to do anything to tailor the noise of the input circuit.

See section 2.1.3 of https://arxiv.org/pdf/1509.02921 If a Pauli Transfer Matrix of some noise channel has off-diagonal terms, we expect the noise to be tailored after twirling.

Author: I did change the noise to a coherent one, please take a look.

state. See the [Executors](executors.md) section for more information on
how to define more advanced executors.

During execution by the simulator, a coherent error is introduced in the circuit

> by applying a rotation around the X-axis (Rx gate) to each output of any 2-qubit gate.

Reviewer 2: ```suggestion
by applying a rotation around the X-axis (Rx gate) to each output of any 2-qubit gate in the circuit of interest.
```

Author: ok, but then I am removing the word circuit from the line above

This noise model is well-suited to highlight the effect of Pauli Twirling,

> which is a technique that turns coherent noise into incoherent Pauli channels.

Reviewer 2: ```suggestion
which is a technique that transforms coherent noise into incoherent noise. The modified noise channel is described using Paulis.
```

> For the sake of this example, we define the noise level as the angle of the Rx rotation.

Reviewer 2: ```suggestion
For the sake of this example executed by a simulator, we define the noise level as an Rx rotation by some angle.
```

Move this line to before L55

Author: I moved this line up, and changed the meaning of noise_level, take a look at commit 37258d62b113ea2513e124b0acb56055ce421d96

Reviewer 2: LGTM! Feel free to merge.

```{code-cell} ipython3
from cirq import CircuitOperation, CXPowGate, CZPowGate, DensityMatrixSimulator, Rx
from cirq.devices.noise_model import GateSubstitutionNoiseModel

def get_noise_model(x_rotation: float) -> GateSubstitutionNoiseModel:

>     """Substitute each C-Z and C-NOT gate in the circuit 

Reviewer 2: ```suggestion
    """Substitute each CZ and CNOT gate in the circuit
```

>     with the gate itself followed by an Rx rotation on the output qubits

Reviewer 2: ```suggestion
    with the gate itself followed by an Rx rotation on the output qubits.
```

    """
    def noisy_c_gate(op):
        if isinstance(op.gate, (CZPowGate, CXPowGate)):
            return CircuitOperation(
                Circuit(
                    op.gate.on(*op.qubits), 
                    Rx(rads=x_rotation).on_each(op.qubits),
                ).freeze())
        return op

    return GateSubstitutionNoiseModel(noisy_c_gate)

def execute(circuit: Circuit, noise_level: float):

>     """Returns Tr[ρ |0⟩⟨0|] where ρ is the state prepared by the circuit"""

Reviewer 2: ```suggestion
    """Returns Tr[ρ |0⟩⟨0|] where ρ is the state prepared by the circuit."""
```

    return (
        DensityMatrixSimulator(noise=get_noise_model(x_rotation=noise_level))
        .simulate(circuit)
        .final_density_matrix[0, 0]
        .real
    )
```

The [executor](executors.md) can be used to evaluate noisy (unmitigated)
expectation values.

```{code-cell} ipython3
# Set the intensity of the noise
NOISE_LEVEL = 0.1

# Compute the expectation value of the |0><0| observable
# in both the noiseless and the noisy setup
ideal_value = execute(circuit, noise_level=0.0)
noisy_value = execute(circuit, noise_level=NOISE_LEVEL)

print(f"Error without twirling: {abs(ideal_value - noisy_value) :.3}")
```

## Apply PT
PT can be applied by first generating twirled variants of the circuit with the function
{func}`.generate_pauli_twirl_variants` from the `mitiq.pt` module, 

> and then averaging over the results obtained by executing those variants.

Reviewer 2: You should make it clear here that you are using the default value of `num_circuits = 10` in `generate_pauli_twirl_variants`.The more variants are generated and averaged over, the better the results are.

from cirq import CircuitOperation, CXPowGate, CZPowGate, DensityMatrixSimulator, Rx
from cirq.devices.noise_model import GateSubstitutionNoiseModel

def get_noise_model(noise_level: float) -> GateSubstitutionNoiseModel:
    """Substitute each CZ and CNOT gate in the circuit
    with the gate itself followed by an Rx rotation on the output qubits.
    """
    rads = pi / 2 * noise_level
    def noisy_c_gate(op):
        if isinstance(op.gate, (CZPowGate, CXPowGate)):
            return CircuitOperation(
                Circuit(
                    op.gate.on(*op.qubits), 
                    Rx(rads=rads).on_each(op.qubits),
                ).freeze())
        return op

    return GateSubstitutionNoiseModel(noisy_c_gate)

def execute(circuit: Circuit, noise_level: float):
    """Returns Tr[ρ |0⟩⟨0|] where ρ is the state prepared by the circuit."""
    return (
        DensityMatrixSimulator(noise=get_noise_model(noise_level=noise_level))
        .simulate(circuit)
        .final_density_matrix[0, 0]
        .real
    )
```

The [executor](executors.md) can be used to evaluate noisy (unmitigated)
expectation values.

```{code-cell} ipython3
# Set the intensity of the noise

> NOISE_LEVEL = 0.1

Reviewer: Good idea setting this explicitly to avoid relying on default values that are easy to miss when reading.

# Compute the expectation value of the |0><0| observable
# in both the noiseless and the noisy setup
ideal_value = execute(circuit, noise_level=0.0)
noisy_value = execute(circuit, noise_level=NOISE_LEVEL)

print(f"Error without twirling: {abs(ideal_value - noisy_value) :.3}")
```

## Apply PT
PT can be applied by first generating twirled variants of the circuit with the function
{func}`.generate_pauli_twirl_variants` from the `mitiq.pt` module, 
and then averaging over the results obtained by executing those variants.

The more variants are generated and averaged over, the more visible the results of PT are.
In this example we generate 5 twirled variants of the circuit, by setting the  `num_circuits` 
argument of the function {func}`.generate_pauli_twirl_variants` (default value is 10.)

```{code-cell} ipython3
from functools import partial
import numpy as np

> from mitiq.executor.executor import Executor

Reviewer: `Executor` is available at the top-level:
```suggestion
from mitiq import Executor
```

from mitiq.pt import generate_pauli_twirl_variants

# Generate twirled circuits
NUM_TWIRLED_VARIANTS = 5
twirled_circuits = generate_pauli_twirl_variants(circuit, num_circuits=NUM_TWIRLED_VARIANTS)
# Average results executed over twirled circuits
pt_vals = Executor(partial(execute, noise_level=NOISE_LEVEL)).evaluate(twirled_circuits)
mitigated_result = np.average(pt_vals)

print(f"Error with twirling: {abs(ideal_value - mitigated_result) :.3}")
```

The idea behind Pauli Twirling is that it leaves the effective logical circuit unchanged, while tailoring the noise into stochastic Pauli errors.

```{admonition} Note:
Pauli Twirling is designed to transform noise, such as the coherent noise simulated in the example above,
but it should not be expected to always have a positive effect. In this sense, it is more of a noise tailoring technique, 

> designed to be composed with other techniques rather than an error mitigation technique in itself.

Reviewer: Can we add a note here that one goal of PT is to make other techniques more effective when stacked together?
